File: backend/app/services/geocoding_service.py
import requests


class GeocodingService:

    def get_coordinates(
        self,
        village: str,
        district: str,
        state: str,
        country: str
    ):

        address = f"{village}, {district}, {state}, {country}"

        url = "https://nominatim.openstreetmap.org/search"

        params = {
            "q": address,
            "format": "json",
            "limit": 1
        }

        headers = {
            "User-Agent": "SolarWindDeploymentPlatform/1.0"
        }

        response = requests.get(
            url,
            params=params,
            headers=headers,
            timeout=10
        )

        response.raise_for_status()

        data = response.json()

        if not data:
            raise ValueError(
                "Location could not be found"
            )

        return {
            "latitude": float(data[0]["lat"]),
            "longitude": float(data[0]["lon"]),
            "display_name": data[0]["display_name"]
        }

File: frontend/src/services/geocodingService.js
export async function geocodeLocation({
  village,
  district,
  state,
  country
}) {
  const query = [
    village,
    district,
    state,
    country
  ]
    .filter(Boolean)
    .join(", ");

  console.log("Searching:", query);

  const url =
    `https://nominatim.openstreetmap.org/search` +
    `?format=json` +
    `&q=${encodeURIComponent(query)}` +
    `&limit=5` +
    `&addressdetails=1`;

  try {
    const response = await fetch(url, {
      headers: {
        Accept: "application/json"
      }
    });

    if (!response.ok) {
      throw new Error(`Geocoding HTTP error: ${response.status}`);
    }

    const results = await response.json();

    console.log("Geocoding results:", results);

    if (!results || results.length === 0) {
      return null;
    }

    const result = results[0];

    return {
      latitude: Number(result.lat),
      longitude: Number(result.lon),
      displayName: result.display_name
    };

  } catch (error) {
    console.error("Geocoding error:", error);
    throw error;
  }
}